README.md
# twitter-card
[![crates.io version][1]][2] [![build status][3]][4]
[![downloads][5]][6] [![docs.rs docs][7]][8]

Generate HTML for Twitter Card integration.

- [Documentation][8]
- [Crates.io][2]
- [Releases][releases]

## Examples
__Basic usage__
```rust
use twitter_card::{Summary, TwitterCard};

let card = Summary::builder()
  .site("@flickr")
  .title("Small Island Developing States Photo Submission")
  .desc("View the album on Flickr.")
  .image("https://farm6.staticflickr.com/5510/14338202952_93595258ff_z.jpg")
  .build();
```

```html
<--! Output -->
<meta name="twitter:card" content="summary" />
<meta name="twitter:site" content="@flickr" />
<meta name="twitter:title" content="Small Island Developing States Photo Submission" />
<meta name="twitter:description" content="View the album on Flickr." />
<meta name="twitter:image" content="https://farm6.staticflickr.com/5510/14338202952_93595258ff_z.jpg" />
```

## Installation
```sh
$ cargo add twitter-card
```

## Safety
This crate uses ``#![deny(unsafe_code)]`` to ensure everything is implemented in
100% Safe Rust.

## Contributing
Want to join us? Check out our ["Contributing" guide][contributing] and take a
look at some of these issues:

- [Issues labeled "good first issue"][good-first-issue]
- [Issues labeled "help wanted"][help-wanted]

## References
- [Twitter Card Validator](https://cards-dev.twitter.com/validator)
- [Twitter Card Overview](https://developer.twitter.com/en/docs/tweets/optimize-with-cards/overview/markup)

## License
[MIT](./LICENSE-MIT) OR [Apache-2.0](./LICENSE-APACHE)

[1]: https://img.shields.io/crates/v/twitter-card.svg?style=flat-square
[2]: https://crates.io/crates/twitter-card
[3]: https://img.shields.io/travis/rust-net-web/twitter-card/master.svg?style=flat-square
[4]: https://travis-ci.org/rust-net-web/twitter-card
[5]: https://img.shields.io/crates/d/twitter-card.svg?style=flat-square
[6]: https://crates.io/crates/twitter-card
[7]: https://img.shields.io/badge/docs-latest-blue.svg?style=flat-square
[8]: https://docs.rs/twitter-card

[releases]: https://github.com/rust-net-web/twitter-card/releases
[contributing]: https://github.com/rust-net-web/twitter-card/blob/master.github/CONTRIBUTING.md
[good-first-issue]: https://github.com/rust-net-web/twitter-card/labels/good%20first%20issue
[help-wanted]: https://github.com/rust-net-web/twitter-card/labels/help%20wanted


CHANGELOG.md
## 2018-11-18, Version 1.0.1
### Commits
- [[`b04dd4aefb`](https://github.com/rust-net-web/twitter-card/commit/b04dd4aefb3eebc4a5cfa65f0367862b38611f2c)] (cargo-release) version 1.0.1 (Yoshua Wuyts)
- [[`64667356f4`](https://github.com/rust-net-web/twitter-card/commit/64667356f4e65a57fe508e0d65e8768e4d85f7bb)] fix missing apis (Yoshua Wuyts)
- [[`040b4a0c97`](https://github.com/rust-net-web/twitter-card/commit/040b4a0c97f04655da13c15b6c0a274979cf9c5a)] cargo fmt (Yoshua Wuyts)
- [[`616ca0c980`](https://github.com/rust-net-web/twitter-card/commit/616ca0c9808cbb0369dd9f3c5c1a2b187ac43580)] Fix typo (#1) (Jan-Erik Rediger)
- [[`8989f074b9`](https://github.com/rust-net-web/twitter-card/commit/8989f074b9768bbd0f348e126e493edb424a5ec1)] Update changelog (Yoshua Wuyts)

### Stats
```diff
 CHANGELOG.md  | 38 ++++++++++++++++++++++++++++++++++++++
 Cargo.toml    |  2 +-
 src/app.rs    |  2 +-
 src/lib.rs    |  6 ++++++
 src/player.rs |  6 +++---
 tests/test.rs |  2 +-
 6 files changed, 50 insertions(+), 6 deletions(-)
```


## 2018-11-17, Version 1.0.0
### Commits
- [[`62da38ba71`](https://github.com/rust-net-web/twitter-card/commit/62da38ba71a679a5ab799216aa66a25cc97595a5)] (cargo-release) version 1.0.0 (Yoshua Wuyts)
- [[`19eabf0fae`](https://github.com/rust-net-web/twitter-card/commit/19eabf0faefd25edfb81ba622793907975aa54f7)] app card + finalize (Yoshua Wuyts)
- [[`cb8cb958fe`](https://github.com/rust-net-web/twitter-card/commit/cb8cb958feed5994370896b61233d0eae3d4a5ff)] player (Yoshua Wuyts)
- [[`5a9a45993e`](https://github.com/rust-net-web/twitter-card/commit/5a9a45993e2d3b3dfe00df38a518f6db66fb57d4)] summary large image (Yoshua Wuyts)
- [[`7ac8b97875`](https://github.com/rust-net-web/twitter-card/commit/7ac8b97875c766b8d858c696bace36815fd70709)] init (Yoshua Wuyts)

### Stats
```diff
 .github/CODE_OF_CONDUCT.md                |  75 ++++++++-
 .github/CONTRIBUTING.md                   |  63 ++++++-
 .github/ISSUE_TEMPLATE.md                 |   9 +-
 .github/ISSUE_TEMPLATE/bug_report.md      |  23 ++-
 .github/ISSUE_TEMPLATE/feature_request.md |  43 ++++-
 .github/ISSUE_TEMPLATE/question.md        |  18 ++-
 .github/PULL_REQUEST_TEMPLATE.md          |  21 ++-
 .github/stale.yml                         |  17 ++-
 .gitignore                                |   7 +-
 .travis.yml                               |  13 +-
 CERTIFICATE                               |  37 ++++-
 Cargo.toml                                |  16 ++-
 LICENSE-APACHE                            | 190 ++++++++++++++++++++-
 LICENSE-MIT                               |  21 ++-
 README.md                                 |  68 +++++++-
 examples/example.rs                       |  14 +-
 rustfmt.toml                              |   2 +-
 src/app.rs                                | 121 ++++++++++++-
 src/error.rs                              |  73 ++++++++-
 src/lib.rs                                | 301 +++++++++++++++++++++++++++++++-
 src/player.rs                             | 118 ++++++++++++-
 src/summary.rs                            |  85 +++++++++-
 src/summary_large_image.rs                |  92 +++++++++-
 tests/test.rs                             |   9 +-
 24 files changed, 1436 insertions(+)
```




tests/test.rs
extern crate failure;
extern crate twitter_card;

use failure::Error;

#[test]
fn should_work() -> Result<(), Error> {
  Ok(())
}


src/app.rs
use super::*;

/// Create an app card.
///
/// The App Card is a great way to represent mobile applications on Twitter and
/// to drive installs. We designed the App Card to allow for a name, description
/// and icon, and also to highlight attributes such as the rating and the price.
///
/// [Read more](https://developer.twitter.com/en/docs/tweets/optimize-with-cards/overview/app-card)
#[derive(Debug, Clone)]
pub struct App {
  strings: Vec<String>,
}

impl App {
  /// Create a new instance.
  pub fn builder() -> Self {
    Self {
      strings: vec![create_card("summary")],
    }
  }

  /// @username of website. Either twitter:site or twitter:site:id is required.
  #[inline]
  pub fn site(mut self, site: &str) -> Self {
    self.strings.push(create_site(site));
    self
  }

  /// Same as twitter:site, but the user’s Twitter ID. Either twitter:site or
  /// twitter:site:id is required.
  #[inline]
  pub fn site_id(mut self, content: &str) -> Self {
    self.strings.push(create_site_id(content));
    self
  }

  /// Description of content.
  ///
  /// ## Panics.
  /// Panics if the description is more than 200 characters.
  #[inline]
  pub fn desc(mut self, content: &str) -> Self {
    self.strings.push(create_desc(content));
    self
  }

  /// Your app ID in the iTunes App Store (Note: NOT your bundle ID).
  #[inline]
  pub fn app_id_iphone(mut self, content: &str) -> Self {
    self.strings.push(create_app_id_iphone(content));
    self
  }

  /// Name of your iPhone app.
  #[inline]
  pub fn app_name_iphone(mut self, content: &str) -> Self {
    self.strings.push(create_app_name_iphone(content));
    self
  }

  /// Your app’s custom URL scheme (you must include ”://” after your scheme
  /// name).
  #[inline]
  pub fn app_url_iphone(mut self, content: &str) -> Self {
    self.strings.push(create_app_url_iphone(content));
    self
  }

  /// Name of your iPad app.
  #[inline]
  pub fn app_name_ipad(mut self, content: &str) -> Self {
    self.strings.push(create_app_name_ipad(content));
    self
  }

  /// Create a Twitter app id iPad tag.
  ///
  /// Your app ID in the iTunes App Store (Note: NOT your bundle ID).
  #[inline]
  pub fn app_id_ipad(mut self, content: &str) -> Self {
    self.strings.push(create_app_id_ipad(content));
    self
  }

  /// Your app’s custom URL scheme (you must include ”://” after your scheme
  /// name).
  #[inline]
  pub fn app_url_ipad(mut self, content: &str) -> Self {
    self.strings.push(create_app_url_ipad(content));
    self
  }

  /// Name of your Android app.
  #[inline]
  pub fn app_name_googleplay(mut self, content: &str) -> Self {
    self.strings.push(create_app_name_googleplay(content));
    self
  }

  /// Your app ID in the Google Play Store.
  #[inline]
  pub fn app_id_googleplay(mut self, content: &str) -> Self {
    self.strings.push(create_app_id_googleplay(content));
    self
  }

  /// Your app’s custom URL scheme.
  #[inline]
  pub fn app_url_googleplay(mut self, content: &str) -> Self {
    self.strings.push(create_app_url_googleplay(content));
    self
  }
}

impl TwitterCard for App {
  #[inline]
  fn build(self) -> String {
    self.strings.join("\n")
  }
}


src/player.rs
use super::*;

/// Create a player card.
///
/// Player cards allow rich media to be embedded on Twitter. This includes, but
/// is not limited to, audio clips and videos.
///
/// [Read more](https://developer.twitter.com/en/docs/tweets/optimize-with-cards/overview/player-card)
#[derive(Debug, Clone)]
pub struct Player {
  strings: Vec<String>,
}

impl Player {
  /// Create a new instance.
  pub fn builder() -> Self {
    Self {
      strings: vec![create_card("summary")],
    }
  }

  /// Title of content.
  ///
  /// ## Panics.
  /// Panics if the description is more than 70 characters.
  #[inline]
  pub fn title(mut self, content: &str) -> Self {
    self.strings.push(create_title(content));
    self
  }

  /// @username of website. Either twitter:site or twitter:site:id is required.
  #[inline]
  pub fn site(mut self, site: &str) -> Self {
    self.strings.push(create_site(site));
    self
  }

  /// Same as twitter:site, but the user’s Twitter ID. Either twitter:site or
  /// twitter:site:id is required.
  #[inline]
  pub fn site_id(mut self, content: &str) -> Self {
    self.strings.push(create_site_id(content));
    self
  }

  /// Twitter user ID of content creator.
  #[inline]
  pub fn creator_id(mut self, content: &str) -> Self {
    self.strings.push(create_creator_id(content));
    self
  }

  /// Description of content.
  ///
  /// ## Panics.
  /// Panics if the description is more than 200 characters.
  #[inline]
  pub fn desc(mut self, content: &str) -> Self {
    self.strings.push(create_desc(content));
    self
  }

  /// URL of image to use in the card. Images must be less than 5MB in size.
  /// JPG, PNG, WEBP and GIF formats are supported. Only the first frame of an
  /// animated GIF will be used. SVG is not supported.
  #[inline]
  pub fn image(mut self, content: &str) -> Self {
    self.strings.push(create_image(content));
    self
  }

  /// A text description of the image conveying the essential nature of an image
  /// to users who are visually impaired. Maximum 420 characters.
  ///
  /// ## Panics.
  /// Panics if the description is more than 420 characters.
  #[inline]
  pub fn image_alt(mut self, content: &str) -> Self {
    self.strings.push(create_image_alt(content));
    self
  }

  /// HTTPS URL of player iframe
  #[inline]
  pub fn player(mut self, content: &str) -> Self {
    self.strings.push(create_player(content));
    self
  }

  /// Width of iframe in pixels.
  #[inline]
  pub fn player_width(mut self, content: &str) -> Self {
    self.strings.push(create_player_width(content));
    self
  }

  /// Height of iframe in pixels.
  #[inline]
  pub fn player_height(mut self, content: &str) -> Self {
    self.strings.push(create_player_height(content));
    self
  }

  /// URL to raw video or audio stream.
  #[inline]
  pub fn player_stream(mut self, content: &str) -> Self {
    self.strings.push(create_player_stream(content));
    self
  }
}

impl TwitterCard for Player {
  #[inline]
  fn build(self) -> String {
    self.strings.join("\n")
  }
}


src/summary.rs
use super::*;

/// Create a summary card.
#[derive(Debug, Clone)]
pub struct Summary {
  strings: Vec<String>,
}

impl Summary {
  /// Create a new instance.
  pub fn builder() -> Self {
    Self {
      strings: vec![create_card("summary")],
    }
  }

  /// @username of website. Either twitter:site or twitter:site:id is required.
  #[inline]
  pub fn site(mut self, site: &str) -> Self {
    self.strings.push(create_site(site));
    self
  }

  /// Same as twitter:site, but the user’s Twitter ID. Either twitter:site or
  /// twitter:site:id is required.
  #[inline]
  pub fn site_id(mut self, content: &str) -> Self {
    self.strings.push(create_site_id(content));
    self
  }

  /// Description of content.
  ///
  /// ## Panics.
  /// Panics if the description is more than 200 characters.
  #[inline]
  pub fn desc(mut self, content: &str) -> Self {
    self.strings.push(create_desc(content));
    self
  }

  /// Title of content.
  ///
  /// ## Panics.
  /// Panics if the description is more than 70 characters.
  #[inline]
  pub fn title(mut self, content: &str) -> Self {
    self.strings.push(create_title(content));
    self
  }

  /// URL of image to use in the card. Images must be less than 5MB in size. JPG,
  /// PNG, WEBP and GIF formats are supported. Only the first frame of an animated
  /// GIF will be used. SVG is not supported.
  #[inline]
  pub fn image(mut self, content: &str) -> Self {
    self.strings.push(create_image(content));
    self
  }

  /// A text description of the image conveying the essential nature of an image
  /// to users who are visually impaired. Maximum 420 characters.
  ///
  /// ## Panics.
  /// Panics if the description is more than 420 characters.
  #[inline]
  pub fn image_alt(mut self, content: &str) -> Self {
    self.strings.push(create_image_alt(content));
    self
  }
}

impl TwitterCard for Summary {
  #[inline]
  fn build(self) -> String {
    self.strings.join("\n")
  }
}


src/summary_large_image.rs
use super::*;

/// Create a summary with large image card.
#[derive(Debug, Clone)]
pub struct SummaryLargeImage {
  strings: Vec<String>,
}

impl SummaryLargeImage {
  /// Create a new instance.
  pub fn builder() -> Self {
    Self {
      strings: vec![create_card("summary_large_image")],
    }
  }

  /// @username of website. Either twitter:site or twitter:site:id is required.
  #[inline]
  pub fn site(mut self, site: &str) -> Self {
    self.strings.push(create_site(site));
    self
  }

  /// Same as twitter:site, but the user’s Twitter ID. Either twitter:site or
  /// twitter:site:id is required.
  #[inline]
  pub fn site_id(mut self, content: &str) -> Self {
    self.strings.push(create_site_id(content));
    self
  }

  /// Twitter user ID of content creator.
  #[inline]
  pub fn creator_id(mut self, content: &str) -> Self {
    self.strings.push(create_creator_id(content));
    self
  }

  /// @username of content creator.
  #[inline]
  pub fn creator(mut self, content: &str) -> Self {
    self.strings.push(create_creator(content));
    self
  }

  /// Description of content.
  ///
  /// ## Panics.
  /// Panics if the description is more than 200 characters.
  #[inline]
  pub fn desc(mut self, content: &str) -> Self {
    self.strings.push(create_desc(content));
    self
  }

  /// Title of content.
  ///
  /// ## Panics.
  /// Panics if the description is more than 70 characters.
  #[inline]
  pub fn title(mut self, content: &str) -> Self {
    self.strings.push(create_title(content));
    self
  }

  /// URL of image to use in the card. Images must be less than 5MB in size.
  /// JPG, PNG, WEBP and GIF formats are supported. Only the first frame of an
  /// animated GIF will be used. SVG is not supported.
  #[inline]
  pub fn image(mut self, content: &str) -> Self {
    self.strings.push(create_image(content));
    self
  }

  /// A text description of the image conveying the essential nature of an image
  /// to users who are visually impaired. Maximum 420 characters.
  ///
  /// ## Panics.
  /// Panics if the description is more than 420 characters.
  #[inline]
  pub fn image_alt(mut self, content: &str) -> Self {
    self.strings.push(create_image_alt(content));
    self
  }
}

impl TwitterCard for SummaryLargeImage {
  #[inline]
  fn build(self) -> String {
    self.strings.join("\n")
  }
}


src/lib.rs
#![forbid(unsafe_code, future_incompatible)]
#![forbid(rust_2018_idioms, rust_2018_compatibility)]
#![deny(missing_debug_implementations, bad_style)]
#![deny(missing_docs)]
#![cfg_attr(test, deny(warnings))]

//! Generate HTML for Twitter Card integration.
//!
//! ## Example
//! ```rust
//! use twitter_card::{Summary, TwitterCard};
//!
//! let card = Summary::builder()
//!   .site("@flickr")
//!   .title("Small Island Developing States Photo Submission")
//!   .desc("View the album on Flickr.")
//!   .image("https://farm6.staticflickr.com/5510/14338202952_93595258ff_z.jpg")
//!   .build();
//! ```
//!
//! ```html
//! <meta name="twitter:card" content="summary" />
//! <meta name="twitter:site" content="@flickr" />
//! <meta name="twitter:title" content="Small Island Developing States Photo Submission" />
//! <meta name="twitter:description" content="View the album on Flickr." />
//! <meta name="twitter:image" content="https://farm6.staticflickr.com/5510/14338202952_93595258ff_z.jpg" />
//! ```

mod app;
mod player;
mod summary;
mod summary_large_image;

pub use crate::app::App;
pub use crate::player::Player;
pub use crate::summary::Summary;
pub use crate::summary_large_image::SummaryLargeImage;

/// Twitter Card.
pub trait TwitterCard {
  /// Convert the Twitter Card to a string.
  fn build(self) -> String;
}

/// Create a Twitter card tag.
///
/// The card type.
///
/// _Used with all cards._
#[inline]
pub fn create_card(content: &str) -> String {
  create("twitter:card", content)
}

/// Create a Twitter site tag.
///
/// @username of website. Either twitter:site or twitter:site:id is required.
///
/// _Used with summary, summary_large_image, app, player cards._
#[inline]
pub fn create_site(content: &str) -> String {
  create("twitter:site", content)
}

/// Create a Twitter site id tag.
///
/// Same as twitter:site, but the user’s Twitter ID. Either twitter:site or
/// twitter:site:id is required.
///
/// _Used with summary, summary_large_image, player cards_
#[inline]
pub fn create_site_id(content: &str) -> String {
  create("twitter:site:id", content)
}

/// Create a Twitter description tag.
///
/// Description of content.
///
/// _Used with summary, summary_large_image, player cards._
///
/// ## Panics.
/// Panics if the description is more than 200 characters.
#[inline]
pub fn create_desc(content: &str) -> String {
  debug_assert!(
    content.len() <= 200,
    "Description has a maximum of 200 characters"
  );
  create("twitter:description", content)
}

/// Create a Twitter title tag.
///
/// Title of content.
///
/// _Used with summary, summary_large_image, player cards._
///
/// ## Panics.
/// Panics if the description is more than 70 characters.
#[inline]
pub fn create_title(content: &str) -> String {
  debug_assert!(content.len() <= 70, "Title has a maximum of 70 characters");
  create("twitter:title", content)
}

/// Create a Twitter image tag.
///
/// URL of image to use in the card. Images must be less than 5MB in size. JPG,
/// PNG, WEBP and GIF formats are supported. Only the first frame of an animated
/// GIF will be used. SVG is not supported.
///
/// _Used with summary, summary_large_image, player cards._
///
#[inline]
pub fn create_image(content: &str) -> String {
  create("twitter:image", content)
}

/// Create a Twitter image alt tag.
///
/// A text description of the image conveying the essential nature of an image
/// to users who are visually impaired. Maximum 420 characters.
///
/// _Used with summary, summary_large_image, player cards._
///
/// ## Panics.
/// Panics if the description is more than 420 characters.
#[inline]
pub fn create_image_alt(content: &str) -> String {
  debug_assert!(
    content.len() <= 420,
    "Image alt has a maximum of 420 characters"
  );
  create("twitter:image:alt", content)
}

/// Create a Twitter creator id tag.
///
/// Twitter user ID of content creator.
///
/// _Used with summary, summary_large_image cards_
#[inline]
pub fn create_creator_id(content: &str) -> String {
  create("twitter:creator:id", content)
}

/// Create a Twitter creator tag.
///
/// @username of content creator.
///
/// _Used with summary_large_image cards._
#[inline]
pub fn create_creator(content: &str) -> String {
  create("twitter:creator", content)
}

/// Create a Twitter player tag.
///
/// HTTPS URL of player iframe
///
/// _Used with player card_
///
#[inline]
pub fn create_player(content: &str) -> String {
  create("twitter:player", content)
}

/// Create a Twitter player width tag.
///
/// Width of iframe in pixels.
///
/// _Used with player card_
///
#[inline]
pub fn create_player_width(content: &str) -> String {
  create("twitter:player:width", content)
}

/// Create a Twitter player height tag.
///
/// Height of iframe in pixels.
///
/// _Used with player card_
///
#[inline]
pub fn create_player_height(content: &str) -> String {
  create("twitter:player:height", content)
}

/// Create a Twitter player stream tag.
///
/// URL to raw video or audio stream.
///
/// _Used with player card_
///
#[inline]
pub fn create_player_stream(content: &str) -> String {
  create("twitter:player:stream", content)
}

/// Create a Twitter app name iPhone tag.
///
/// Name of your iPhone app.
///
/// _Used with app card_
///
#[inline]
pub fn create_app_name_iphone(content: &str) -> String {
  create("twitter:app:name:iphone", content)
}

/// Create a Twitter app id iPhone tag.
///
/// Your app ID in the iTunes App Store (Note: NOT your bundle ID).
///
/// _Used with app card_
///
#[inline]
pub fn create_app_id_iphone(content: &str) -> String {
  create("twitter:app:id:iphone", content)
}

/// Create a Twitter app url iPhone tag.
///
/// Your app’s custom URL scheme (you must include ”://” after your scheme
/// name).
///
/// _Used with app card_
///
#[inline]
pub fn create_app_url_iphone(content: &str) -> String {
  create("twitter:app:url:iphone", content)
}

/// Create a Twitter app name iPad tag.
///
/// Name of your iPad app.
///
/// _Used with app card_
///
#[inline]
pub fn create_app_name_ipad(content: &str) -> String {
  create("twitter:app:name:ipad", content)
}

/// Create a Twitter app id iPad tag.
///
/// Your app ID in the iTunes App Store (Note: NOT your bundle ID).
///
/// _Used with app card_
///
#[inline]
pub fn create_app_id_ipad(content: &str) -> String {
  create("twitter:app:id:ipad", content)
}

/// Create a Twitter app url iPad tag.
///
/// Your app’s custom URL scheme (you must include ”://” after your scheme
/// name).
///
/// _Used with app card_
///
#[inline]
pub fn create_app_url_ipad(content: &str) -> String {
  create("twitter:app:url:ipad", content)
}

/// Create a Twitter app name Google Play tag.
///
/// Name of your Android app.
///
/// _Used with app card_
///
#[inline]
pub fn create_app_name_googleplay(content: &str) -> String {
  create("twitter:app:name:googleplay", content)
}

/// Create a Twitter app id Google Play tag.
///
/// Your app ID in the Google Play Store.
///
/// _Used with app card_
///
#[inline]
pub fn create_app_id_googleplay(content: &str) -> String {
  create("twitter:app:id:googleplay", content)
}

/// Create a Twitter app url Google Play tag.
///
/// Your app’s custom URL scheme.
///
/// _Used with app card_
///
#[inline]
pub fn create_app_url_googleplay(content: &str) -> String {
  create("twitter:app:url:googleplay", content)
}

/// Create an HTML tag
#[inline]
fn create(name: &str, content: &str) -> String {
  format!(r#"<meta name="{}" content="{}" />"#, name, content)
}
